api/gcp/firestore/firestore.go

```diff
@@ -60,39 +174,39 @@ func getByDoc(ctx context.Context, doc *firestore.DocumentRef) (*FileMeta, error
 
 	snapshot, err := doc.Get(ctx)
 	if err != nil {
-		log.Println("Fail to get document from firestore:", err)
+		log.Printf("failed to get document from firestore: %v", err)
 		return nil, err
 	}
-	result, err := toResult(snapshot)
+	result, err := newFileMeta(snapshot)
 	if err != nil {
 		return nil, err
 	}
 	return result, nil
 }
 
-func toResult(snapshot *firestore.DocumentSnapshot) (*FileMeta, error) {
-	result := &FileMeta{}
-	if err := snapshot.DataTo(result); err != nil {
-		log.Println("Fail to format result snapshot:", snapshot, "err:", err)
-		return result, err
+func newFileMeta(snapshot *firestore.DocumentSnapshot) (*FileMeta, error) {
+	data := snapshot.Data()
+
+	var result = new(FileMeta)
+	if err := result.Set(data); err != nil {
+		return result, fmt.Errorf("failed to format result snapshot: %v, err: %w", snapshot, err)
 	}
+
 	result.ID = snapshot.Ref.ID
 	result.CreateTime = snapshot.CreateTime
 	result.UpdateTime = snapshot.UpdateTime
 	return result, nil
 }
 
-// ListByTags lists the FileMeta from given tags
-func ListByTags(ctx context.Context, client *firestore.Client, tags []string, orderNo string, size int) ([]*FileMeta, error) {
+// ListByTags lists the FileMeta from given tags.
+func (c *firestoreClient) ListByTags(ctx context.Context, tags []string, orderNo string, size int) ([]*FileMeta, error) {
 	ctx, cancel := context.WithTimeout(ctx, timeout)
```

Comment:
issue: I haven't yet seen where the Firestore client is closed when we're doing using it.

Reply:
The implementation is in the "api/files/routes.go". 
At line 375, it uses "defer dbClient.Close()" to close the client

https://github.com/GoogleCloudPlatform/app-large-data-sharing-golang/pull/34/commits/2c1b6fd19c95db5b7b6714c4611a79f4f617047e#diff-b82dd01612c1af92a7b19d042729a47ce61d4c0d49bee2ec55befa38b7beb2a9R375